File: shortest_path/dijikstra.c
#include <cstdio>
#include <vector>
#include <algorithm>

using namespace std;

const int MAX = 100;

vector <int> graph[MAX];
vector <int> cost[MAX];

int n, m, start, end;

int main(){
  scanf("%d %d %d %d", &n, &m, &start, &end);

  for(int i=0;i<m;i++){
    int a,b,c;

    scanf("%d %d %D", &a, &b, &c);

    graph[a].push_back(b);
    graph[b].push_back(a);
    cost[a].push_back(c);
    cost[b].push_back(c);

  }

  for(int i=0;i<n;i++) Table[i] = 99999999;
  Table[0] = 0;

  for(int i=0;i<n;i++){
    int minValue = 9999999, minIndex = -1;
    for(int j=0;j<n;j++){
      if(!Check[j] && minValue > Table[j]) {
        minValue = Table[j];
        minIndex = j;
      }
    }
    Check[minIndex] = true;
    for(int j=0;j<graph[minIndex].size();j++){
      int Node2 = graph[minIndex][j];
      int Cost2 = cost[minIndex][j];

      if(Table[Node2]> Table[minIndex] + Cost2){
        Table[Node2] = Table[minIndex] + Cost2;
      }
    }
  }
  printf("%d\n", Table[end2]);
  return 0;
}

File: shortest_path/dijikstra.py
from collections import defaultdict

graph = defaultdict(list)
table = defaultdict(int)
check = defaultdict(bool)

n, m = input().split()
n, m = int(n), int(m)

for i in range(m):
  a, b = input().split()
  a, b = int(a), int(b)
  graph[a].append(b)
  graph[b].append(a)

begin, end = input().split()
begin, end = int(begin), int(end)

for i in range(n):
  table[i] = 999999999
  check[i] = False

table[begin] = 0

# for i = 0 ~ n
for i in range(n):
  minIndex = -1
  minValue = 999999999
  for j in range(n):
    if table[j] <= minValue and not check[j]:
      minValue = table[j]
      minIdex = j

  check[minIdex] = True
  for node in graph[minIdex]:
    if table[node] > table[minIdex] + 1:
      table[node] = table[minIdex] + 1

print(table[end])